File: Services/TimeConversionService.cs
namespace TimezoneConverter.Services;

public sealed record ZoneConversionResult(
    string Abbreviation,
    string WindowsId,
    DateTime LocalWallTime,
    TimeSpan UtcOffset,
    int DayDeltaFromPrimary);

public sealed record ConversionSnapshot(
    DateTime InputWallTime,
    DateTime Utc,
    DateTime PrimaryLocalTime,
    TimeSpan PrimaryUtcOffset,
    IReadOnlyList<ZoneConversionResult> Targets,
    string? Warning = null);

/// <summary>
/// Pure timezone conversion logic (no UI).
/// </summary>
public static class TimeConversionService
{
    public enum LocalTimeKind
    {
        Valid,
        Invalid,   // DST spring-forward gap
        Ambiguous  // DST fall-back overlap
    }

    public static LocalTimeKind ClassifyLocalTime(TimeZoneInfo zone, DateTime wallUnspecified)
    {
        var t = DateTime.SpecifyKind(wallUnspecified, DateTimeKind.Unspecified);
        if (zone.IsInvalidTime(t))
            return LocalTimeKind.Invalid;
        if (zone.IsAmbiguousTime(t))
            return LocalTimeKind.Ambiguous;
        return LocalTimeKind.Valid;
    }

    /// <summary>
    /// Interprets a wall-clock time in <paramref name="inputTimeZone"/> and converts to UTC and targets.
    /// </summary>
    public static ConversionSnapshot Convert(
        DateTime inputWallUnspecified,
        TimeZoneInfo inputTimeZone,
        TimeZoneInfo primaryTimeZone,
        IReadOnlyList<(string Abbreviation, string WindowsId, TimeZoneInfo Zone)> targets)
    {
        var unspecified = DateTime.SpecifyKind(inputWallUnspecified, DateTimeKind.Unspecified);
        string? warning = null;
        var effectiveWall = unspecified;

        if (inputTimeZone.IsInvalidTime(unspecified))
        {
            effectiveWall = ShiftOutOfGap(inputTimeZone, unspecified);
            warning =
                $"{unspecified:h:mm tt} does not exist in {inputTimeZone.Id} (DST spring-forward). " +
                $"Using {effectiveWall:h:mm tt} instead.";
        }

        DateTime utc;
        if (inputTimeZone.IsAmbiguousTime(effectiveWall))
        {
            // The larger offset identifies the earlier occurrence of the repeated wall time.
            var offsets = inputTimeZone.GetAmbiguousTimeOffsets(effectiveWall);
            var preferred = offsets[0] > offsets[1] ? offsets[0] : offsets[1];
            utc = new DateTime(
                Math.Clamp(effectiveWall.Ticks - preferred.Ticks, DateTime.MinValue.Ticks, DateTime.MaxValue.Ticks),
                DateTimeKind.Utc);
            warning ??=
                $"{effectiveWall:h:mm tt} is ambiguous in {inputTimeZone.Id} (DST fall-back). " +
                $"Using the {FormatOffset(preferred)} occurrence.";
        }
        else
        {
            utc = TimeZoneInfo.ConvertTimeToUtc(effectiveWall, inputTimeZone);
        }

        return CreateSnapshot(effectiveWall, utc, primaryTimeZone, targets, warning);
    }

    public static ConversionSnapshot ConvertLiveNow(
        TimeZoneInfo inputTimeZone,
        TimeZoneInfo primaryTimeZone,
        IReadOnlyList<(string Abbreviation, string WindowsId, TimeZoneInfo Zone)> targets)
    {
        var utc = DateTime.UtcNow;
        var inputWall = TimeZoneInfo.ConvertTimeFromUtc(utc, inputTimeZone);
        return CreateSnapshot(inputWall, utc, primaryTimeZone, targets);
    }

    private static DateTime ShiftOutOfGap(TimeZoneInfo zone, DateTime wall)
    {
        // Derive the actual jump: DST gaps can be half an hour or several hours.
        var before = FindOutsideGap(zone, wall, -TimeSpan.TicksPerHour);
        var after = FindOutsideGap(zone, wall, TimeSpan.TicksPerHour);
        if (zone.IsInvalidTime(after))
            return before; // No later wall time is representable at DateTime.MaxValue.

        var gap = zone.GetUtcOffset(after) - zone.GetUtcOffset(before);
        var shifted = AddClamped(wall, gap.Ticks);
        return gap > TimeSpan.Zero && !zone.IsInvalidTime(shifted) ? shifted : after;
    }

    private static DateTime FindOutsideGap(TimeZoneInfo zone, DateTime wall, long stepTicks)
    {
        while (zone.IsInvalidTime(wall))
        {
            var next = AddClamped(wall, stepTicks);
            if (next == wall)
                break;
            wall = next;
        }

        return wall;
    }

    private static DateTime AddClamped(DateTime wall, long ticks) =>
        new(Math.Clamp(wall.Ticks + ticks, DateTime.MinValue.Ticks, DateTime.MaxValue.Ticks), DateTimeKind.Unspecified);

    private static ConversionSnapshot CreateSnapshot(
        DateTime inputWall,
        DateTime utc,
        TimeZoneInfo primaryTimeZone,
        IReadOnlyList<(string Abbreviation, string WindowsId, TimeZoneInfo Zone)> targets,
        string? warning = null)
    {
        var primary = TimeZoneInfo.ConvertTimeFromUtc(utc, primaryTimeZone);
        // An ambiguous wall time loses which occurrence was selected. UTC retains that information.
        var primaryOffset = primaryTimeZone.GetUtcOffset(utc);

        var results = new ZoneConversionResult[targets.Count];
        for (var i = 0; i < targets.Count; i++)
        {
            var (abbr, windowsId, zone) = targets[i];
            var converted = TimeZoneInfo.ConvertTimeFromUtc(utc, zone);
            var offset = zone.GetUtcOffset(utc);
            var dayDelta = (converted.Date - primary.Date).Days;
            results[i] = new ZoneConversionResult(abbr, windowsId, converted, offset, dayDelta);
        }

        return new ConversionSnapshot(inputWall, utc, primary, primaryOffset, results, warning);
    }

    public static string FormatOffset(TimeSpan offset)
    {
        var sign = offset < TimeSpan.Zero ? "-" : "+";
        var abs = offset.Duration();
        return abs.Minutes == 0
            ? $"UTC{sign}{abs.Hours}"
            : $"UTC{sign}{abs.Hours}:{abs.Minutes:D2}";
    }

    public static string FormatDayDelta(int days) =>
        days switch
        {
            0 => string.Empty,
            1 => " +1d",
            -1 => " -1d",
            > 1 => $" +{days}d",
            _ => $" {days}d"
        };

    public static string FormatDigital(DateTime time, bool use24Hour, bool includeSeconds)
    {
        if (includeSeconds)
            return use24Hour ? time.ToString("HH:mm:ss") : time.ToString("hh:mm:ss tt");
        return use24Hour ? time.ToString("HH:mm") : time.ToString("hh:mm tt");
    }

    /// <summary>Multi-line clipboard block.</summary>
    public static string FormatCopyMultiline(ConversionSnapshot snap, bool use24Hour, bool live)
    {
        var lines = new List<string>
        {
            $"ZoneShift ({snap.PrimaryLocalTime:yyyy-MM-dd}{(live ? ", live" : "")})",
            $"Local: {FormatDigital(snap.PrimaryLocalTime, use24Hour, live)} ({FormatOffset(snap.PrimaryUtcOffset)})"
        };
        foreach (var t in snap.Targets)
        {
            var day = FormatDayDelta(t.DayDeltaFromPrimary);
            lines.Add($"{t.Abbreviation}: {FormatDigital(t.LocalWallTime, use24Hour, live)} ({FormatOffset(t.UtcOffset)}{day})");
        }

        if (!string.IsNullOrWhiteSpace(snap.Warning))
            lines.Add($"Note: {snap.Warning}");

        return string.Join(Environment.NewLine, lines);
    }

    /// <summary>Single-line clipboard for chat apps.</summary>
    public static string FormatCopyOneLine(ConversionSnapshot snap, bool use24Hour, bool live)
    {
        var parts = new List<string>
        {
            $"Local {FormatDigital(snap.PrimaryLocalTime, use24Hour, false)}"
        };
        foreach (var t in snap.Targets)
            parts.Add($"{t.Abbreviation} {FormatDigital(t.LocalWallTime, use24Hour, false)}");

        var line = string.Join(" | ", parts);
        if (!string.IsNullOrWhiteSpace(snap.Warning))
            line += " (DST note)";
        return line;
    }
}

File: android/app/src/main/java/com/zoneshift/app/domain/TimeConversionService.kt
package com.zoneshift.app.domain

import java.time.Clock
import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneId
import java.time.ZoneOffset
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit
import java.util.Locale
import kotlin.math.abs

object TimeConversionService {
    fun classifyLocalTime(zoneId: ZoneId, wallTime: LocalDateTime): LocalTimeKind {
        return when (zoneId.rules.getValidOffsets(wallTime).size) {
            0 -> LocalTimeKind.INVALID
            1 -> LocalTimeKind.VALID
            else -> LocalTimeKind.AMBIGUOUS
        }
    }

    fun convert(
        inputWallTime: LocalDateTime,
        inputZoneId: ZoneId,
        primaryZoneId: ZoneId,
        targetZoneIds: List<ZoneId>,
    ): ConversionSnapshot {
        val offsets = inputZoneId.rules.getValidOffsets(inputWallTime)
        val warning: String?
        val resolved: ZonedDateTime

        when (offsets.size) {
            0 -> {
                val transition = requireNotNull(inputZoneId.rules.getTransition(inputWallTime))
                val adjusted = inputWallTime.plus(transition.duration)
                resolved = ZonedDateTime.ofLocal(adjusted, inputZoneId, transition.offsetAfter)
                warning = "${formatWall(inputWallTime)} does not exist in ${friendly(inputZoneId)} " +
                    "because the clocks move forward. Using ${formatWall(adjusted)} instead."
            }

            1 -> {
                resolved = ZonedDateTime.ofLocal(inputWallTime, inputZoneId, offsets.single())
                warning = null
            }

            else -> {
                val preferred = offsets.maxBy { it.totalSeconds }
                resolved = ZonedDateTime.ofLocal(inputWallTime, inputZoneId, preferred)
                warning = "${formatWall(inputWallTime)} occurs twice in ${friendly(inputZoneId)}. " +
                    "Using the ${formatOffset(preferred)} occurrence."
            }
        }

        return snapshot(
            inputWallTime = resolved.toLocalDateTime(),
            instant = resolved.toInstant(),
            sourceZoneId = inputZoneId,
            primaryZoneId = primaryZoneId,
            targetZoneIds = targetZoneIds,
            warning = warning,
        )
    }

    fun convertLiveNow(
        inputZoneId: ZoneId,
        primaryZoneId: ZoneId,
        targetZoneIds: List<ZoneId>,
        clock: Clock = Clock.systemUTC(),
    ): ConversionSnapshot {
        val instant = clock.instant()
        return snapshot(
            inputWallTime = instant.atZone(inputZoneId).toLocalDateTime(),
            instant = instant,
            sourceZoneId = inputZoneId,
            primaryZoneId = primaryZoneId,
            targetZoneIds = targetZoneIds,
            warning = null,
        )
    }

    private fun snapshot(
        inputWallTime: LocalDateTime,
        instant: Instant,
        sourceZoneId: ZoneId,
        primaryZoneId: ZoneId,
        targetZoneIds: List<ZoneId>,
        warning: String?,
    ): ConversionSnapshot {
        val primaryDateTime = instant.atZone(primaryZoneId)
        val primary = result(primaryZoneId, primaryDateTime, primaryDateTime)
        val targets = targetZoneIds.distinct().map { zoneId ->
            result(zoneId, instant.atZone(zoneId), primaryDateTime)
        }
        return ConversionSnapshot(
            inputWallTime = inputWallTime,
            instant = instant,
            sourceZoneId = sourceZoneId,
            primary = primary,
            targets = targets,
            warning = warning,
        )
    }

    private fun result(
        zoneId: ZoneId,
        zonedDateTime: ZonedDateTime,
        primaryDateTime: ZonedDateTime,
    ): ZoneConversionResult {
        val option = ZoneCatalog.optionFor(zoneId.id)
        val abbreviation = DateTimeFormatter.ofPattern("z", Locale.getDefault()).format(zonedDateTime)
        return ZoneConversionResult(
            zoneId = zoneId,
            label = option.label,
            abbreviation = abbreviation,
            localDateTime = zonedDateTime.toLocalDateTime(),
            utcOffset = zonedDateTime.offset,
            dayDeltaFromPrimary = ChronoUnit.DAYS.between(
                primaryDateTime.toLocalDate(),
                zonedDateTime.toLocalDate(),
            ),
        )
    }

    fun formatOffset(offset: ZoneOffset): String {
        val totalMinutes = offset.totalSeconds / 60
        val sign = if (totalMinutes < 0) "-" else "+"
        val absolute = abs(totalMinutes)
        val hours = absolute / 60
        val minutes = absolute % 60
        return if (minutes == 0) "UTC$sign$hours" else "UTC$sign$hours:${minutes.toString().padStart(2, '0')}"
    }

    fun formatDayDelta(days: Long): String = when {
        days == 0L -> ""
        days == 1L -> "Tomorrow"
        days == -1L -> "Yesterday"
        days > 1L -> "+${days} days"
        else -> "${days} days"
    }

    fun formatDigital(time: LocalDateTime, use24Hour: Boolean, includeSeconds: Boolean): String {
        val pattern = when {
            use24Hour && includeSeconds -> "HH:mm:ss"
            use24Hour -> "HH:mm"
            includeSeconds -> "hh:mm:ss a"
            else -> "hh:mm a"
        }
        return time.format(DateTimeFormatter.ofPattern(pattern, Locale.getDefault()))
    }

    fun formatCopyMultiline(snapshot: ConversionSnapshot, use24Hour: Boolean, live: Boolean): String {
        val date = snapshot.primary.localDateTime.toLocalDate()
        val lines = mutableListOf(
            "ZoneShift ($date${if (live) ", live" else ""})",
            "Local: ${formatDigital(snapshot.primary.localDateTime, use24Hour, live)} " +
                "(${formatOffset(snapshot.primary.utcOffset)})",
        )
        snapshot.targets.forEach { target ->
            val day = formatDayDelta(target.dayDeltaFromPrimary)
            val daySuffix = if (day.isEmpty()) "" else ", $day"
            lines += "${target.label}: ${formatDigital(target.localDateTime, use24Hour, live)} " +
                "(${formatOffset(target.utcOffset)}$daySuffix)"
        }
        snapshot.warning?.let { lines += "Note: $it" }
        return lines.joinToString("\n")
    }

    fun formatCopyOneLine(snapshot: ConversionSnapshot, use24Hour: Boolean): String {
        val parts = mutableListOf(
            "Local ${formatDigital(snapshot.primary.localDateTime, use24Hour, false)}"
        )
        snapshot.targets.forEach { target ->
            parts += "${target.label} ${formatDigital(target.localDateTime, use24Hour, false)}"
        }
        return parts.joinToString(" | ") + if (snapshot.warning == null) "" else " (DST note)"
    }

    private fun formatWall(time: LocalDateTime): String =
        time.format(DateTimeFormatter.ofPattern("h:mm a", Locale.getDefault()))

    private fun friendly(zoneId: ZoneId): String = ZoneCatalog.optionFor(zoneId.id).label
}
